pkg/util/awsutil/s3factory/client.go

```diff
@@ -38,8 +41,61 @@ type S3Client struct {
 	configDir string
 }
 
-// NewClientFromSecret returns a S3 client based on given k8s secret containing aws credentials.
-func NewClientFromSecret(kubecli kubernetes.Interface, namespace, endpoint, awsSecret string) (w *S3Client, err error) {
+// ClientConfig is the configuration for a S3 client.
+type ClientConfig struct {
+	Endpoint  string
+	Namespace string
+	AWSSecret string
+}
```

Comment:
Should AWSSecret be more generically something like Auth and direct traffic/validate with for example a switch? Just thinking how we make this more flexible moving forward.

```golang
switch cfg.Auth.Type {
	case "IAMRole":
            return newClientForIAMRole(cfg.Endpoint)
        case "Secret":
           return newClientFromSecret(...
        default:
		// oops  
	}
```

Reply:
I don't know if I understood correctly.

Who would set that information? should the upper layer decide what client type it wants and tell to the factory method what client wants?

(Independently of the type we need `AWSSecret` in case of config base auth to know the name of the secret to reference when downloading and creating an AWS config file)

Reply:
@slok let me rephrase. Since we now create ClientConfig object (which I like) what happens if a 3rd authentication option needs to be added? Currently we only have 2 options and perhaps we should make this more extensible since we are already making this change.

So I was thing something more like this might work.
```
type ClientConfig struct {
	Endpoint  string
	Namespace string
	Auth
}

type Auth struct {
	Type  string
	Value string
}
```
What do you think? Then you can set Auth{Type: "AWSSecret", Value: "Foo"} we need to think about this a bit but I would like to move in this direction in general if possible and I would like to hear your thoughts.

Reply:
Oh! I see I didn't understand at first. I like it, The only caveat that I see here is that depending on the Auth-type could require a different number and kind of fields.

Reply:
Yes this requires thought to do correctly so we should not be in a rush to change this.